spec/compiler/crystal/tools/doc/directives_spec.cr

```diff
@@ -0,0 +1,115 @@
+require "../../../spec_helper"
+
+describe Crystal::Doc::Generator do
+  context ":nodoc:" do
+    it "hides documentation from being generated for methods" do
+      program = top_level_semantic(<<-CRYSTAL, wants_doc: true).program
+        class Foo
+          # :nodoc:
+          #
+          # Some docs
+          def foo
+          end
+        end
+        CRYSTAL
+
+      generator = Doc::Generator.new program, [""]
+      generator.type(program.types["Foo"]).lookup_method("foo").should be_nil
+    end
+
+    it "hides documentation from being generated for classes" do
+      program = top_level_semantic(<<-CRYSTAL, wants_doc: true).program
+        # :nodoc:
+        class Foo
+        end
+        CRYSTAL
+
+      generator = Doc::Generator.new program, [""]
+      generator.must_include?(program.types["Foo"]).should be_false
+    end
+  end
+
+  context ":showdoc:" do
+    it "shows documentation for private methods" do
+      program = top_level_semantic(<<-CRYSTAL, wants_doc: true).program
+        class Foo
+          # :showdoc:
+          #
+          # Some docs
+          private def foo
+          end
+        end
+        CRYSTAL
+
+      generator = Doc::Generator.new program, [""]
+      a_def = generator.type(program.types["Foo"]).lookup_method("foo").not_nil!
+      a_def.doc.should eq("Some docs")
+      a_def.visibility.should eq("private")
+    end
+
+    it "does not include documentation for methods within a :nodoc: namespace" do
+      program = top_level_semantic(<<-CRYSTAL, wants_doc: true).program
+        # :nodoc:
+        class Foo
+          # :showdoc:
+          #
+          # Some docs
+          private def foo
+          end
+        end
+        CRYSTAL
+
+      generator = Doc::Generator.new program, [""]
+
+      # If namespace isn't included, don't need to check if the method is included
+      generator.must_include?(program.types["Foo"]).should be_false
+    end
+
+    it "shows all private and protected methods in a :showdoc: namespace" do
```

Comment:
I'm not sure about this case. It feels like a loaded gun ready to shoot. Does it apply recursively to the methods in a `Foo::Bar::Baz` for example?

I'd prefer each private type or protected/private method to be explicitly marked with the `:showdoc:` directive to keep the behavior change local.

Reply:
That makes sense. My only concern is that it'll be annoying to mark everything inside a `lib` with `:showdoc:` (and it'd be very easy to accidentally miss one). Maybe we could handle those differently though.

Reply:
Ah, `lib` is an excellent counter example!

Reply:
One could argue that the reason `lib` functions and constants don't shown up in the documentation is because `lib` is `:nodoc` by default? but now `:showdoc:` would merely revert that.

Then argue further that it's the same for a private type: marking it `:showdoc:` doesn't mean its private methods should be documented by default, it's only reverting its own `:nodoc:` state?